src/audio/codec_adapter/codec/cadence.c

```diff
@@ -139,6 +139,48 @@ int cadence_codec_init(struct comp_dev *dev)
 	return ret;
 }
 
+static int update_stream_params(struct comp_dev *dev)
+{
+	int ret = 0;
+	struct comp_data *ca_data = comp_get_drvdata(dev);
+	struct codec_data *codec = comp_get_codec(dev);
+	struct cadence_codec_data *cd = codec->private;
+	struct ca_config *ca_config = &ca_data->ca_config;
+	struct sof_ipc_stream_params *stream = &ca_data->stream_params;
+	/* copy stream params localy to ensure container size is correct */
+	uint32_t sample_width = (stream->frame_fmt == SOF_IPC_FRAME_S16_LE) ? 0x10 : 0x20;
+	uint32_t sample_rate = stream->rate;
+	uint32_t channels = stream->channels;
+
+	comp_err(dev, "update_stream_params() start");
+
+	/* update stream parameters */
+	/* sample rate */
+	API_CALL(cd, XA_API_CMD_SET_CONFIG_PARAM, ca_config->sample_rate_id,
+		 &sample_rate, ret);
```

Comment:
This relies on having the ID in the topology and the value in the stream. It doesn't feel good to me.

Reply:
Its a bit awkward I agree but how else would you like to cover stream params update if each codec has a different set of IDs?

Reply:
@mrajwa Like in my code snipet above. We make a function that given a codec id sets the parameter id for our interested parameters. 
I don't think we ever need to setup dynamically the sample rate, channels, width.

For NXP supported Cadence codecs we have the headers released under MIT license so that we can add them to the repo. I even sent a PR for AAC/MP3 for now. 

One problem would stay with your DAP headers. I do think that cadence will release them under MIT if you ask them. Anyhow, for now you can just hardcode the paramter ids for sample rate, channels, width in the translation function.

Also, as an extension I would really like to dynamically setup the codec ID. But that's a thing for the future. For now if you need different codecs you need different topologies, that's not really scalable.

Reply:
I agree. Yes including the header files is problematic due to licence concerns. The other solution would be to just define enums with these three important params like sample_rate sample_width and channels_count. And use the base on codec_id. I think this is your propsal and I like it too

Reply:
@mrajwa yes, you got it right. I hope @lgirdwood can merge this soon https://github.com/thesofproject/sof/pull/4009 so that you can use the defines for sample rate, bitwidth, etc.

Reply:
@dbaluta #4009 now merged.

Reply:
@lgirdwood thanks! @mrajwa I can take over this PR around Tuesday/Wednesday if you are caught with other stuff.